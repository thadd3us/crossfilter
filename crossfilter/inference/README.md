# Crossfilter Inference Tools

This directory contains tools for computing embeddings and running inference on images using various models.

## compute_embeddings_cli.py

A command-line interface for computing embeddings for images and storing them in a SQLite database. The CLI supports incremental computation, UMAP projection, and parallel processing.

### Features

- **Incremental Processing**: Only computes embeddings for images that don't already exist in the database
- **Batch Processing**: Processes images in configurable batches for efficient memory usage
- **Parallel Database Writes**: Uses a background worker thread to write embeddings to the database in parallel with computation
- **UMAP Projection**: Optionally projects embeddings to 2D space using UMAP with spherical geometry
- **Progress Tracking**: Shows progress bars for long-running operations
- **SQLAlchemy Integration**: Uses SQLAlchemy for robust database operations

### Usage

```bash
python -m crossfilter.inference.compute_embeddings_cli \
    --embedding_type SIGLIP2 \
    --input_dir /path/to/images_by_uuid \
    --output_embeddings_db /path/to/output.db \
    --batch_size 10 \
    --recompute_existing_embeddings \
    --reproject_umap_embeddings
```

### Output Database Schema

The CLI creates the following tables:

#### `{EMBEDDING_TYPE}_EMBEDDINGS`
- `UUID` (TEXT PRIMARY KEY): The UUID of the image
- `EMBEDDING` (BLOB): The embedding vector serialized with msgpack

#### `{EMBEDDING_TYPE}_UMAP_MODEL`
- `MODEL` (BLOB): The UMAP model serialized with pickle

