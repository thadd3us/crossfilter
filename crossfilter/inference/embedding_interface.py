"""
Common interface for embedding computation classes.

This module defines the abstract base class that all embedding implementations
should inherit from, ensuring a consistent API across different embedding types.
"""

from abc import ABC, abstractmethod
from pathlib import Path

import numpy as np


class EmbeddingInterface(ABC):
    """Abstract base class for embedding computation."""

    @abstractmethod
    def compute_image_embeddings(self, image_paths: list[Path]) -> list[np.ndarray]:
        """
        Compute embeddings for a list of image paths.

        THAD: TODO: No parallel arrays -- take a DataFrame as input, add a column containing the embeddings.  Don't return anything.

        Raises:
            FileNotFoundError: If any image path doesn't exist
            ValueError: If any image cannot be loaded
        """
        pass

    @abstractmethod
    def compute_text_embeddings(self, captions: list[str]) -> list[np.ndarray]:
        """
        Compute embeddings for a list of text captions.

        Args:
            captions: List of text strings to encode

        Returns:
            List of 1D numpy arrays, one embedding vector per input caption
        """
        pass
