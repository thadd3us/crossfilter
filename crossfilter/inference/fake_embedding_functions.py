"""
Fake embedding functions for testing purposes.

This module provides fake embedding functions that compute simple statistics from images
without requiring heavy ML models or network downloads. The fake embeddings are designed
to be fast, deterministic, and suitable for testing.
"""

import logging
from pathlib import Path

import numpy as np
from PIL import Image

logger = logging.getLogger(__name__)


def compute_image_embeddings(
    image_paths: list[Path],
    batch_size: int = 8,
    model_name: str = "fake-embedding-for-testing",
) -> list[np.ndarray]:
    """
    Compute fake embeddings for a list of image paths by calculating mean and stddev of RGB channels.

    This function provides a fast, deterministic alternative to real embedding models for testing.
    It computes 6-dimensional embeddings: [R_mean, G_mean, B_mean, R_std, G_std, B_std].

    Returns:
        List of 1D numpy arrays, one embedding vector per input image

    Raises:
        FileNotFoundError: If any image path doesn't exist
        ValueError: If any image cannot be loaded
    """
    logger.info(f"Computing fake embeddings for {len(image_paths)} images")

    embeddings = []

    # Process images in batches (for API compatibility)
    # THAD: TODO: Get rid of the baching loop.
    for i in range(0, len(image_paths), batch_size):
        batch_paths = image_paths[i : i + batch_size]

        # Process each image in the batch
        for image_path in batch_paths:
            if not image_path.exists():
                raise FileNotFoundError(f"Image not found: {image_path}")

            try:
                # Load image and convert to RGB if necessary
                image = Image.open(image_path).convert("RGB")

                # Convert to numpy array and normalize to [0, 1]
                image_array = np.array(image).astype(np.float32) / 255.0

                # Calculate mean and standard deviation for each RGB channel
                r_channel = image_array[:, :, 0]
                g_channel = image_array[:, :, 1]
                b_channel = image_array[:, :, 2]

                r_mean = np.mean(r_channel)
                g_mean = np.mean(g_channel)
                b_mean = np.mean(b_channel)

                r_std = np.std(r_channel)
                g_std = np.std(g_channel)
                b_std = np.std(b_channel)

                # Create 6-dimensional embedding: [R_mean, G_mean, B_mean, R_std, G_std, B_std]
                embedding = np.array(
                    [r_mean, g_mean, b_mean, r_std, g_std, b_std], dtype=np.float32
                )

                # Normalize the embedding to unit length (to match SIGLIP2 behavior)
                embedding_norm = np.linalg.norm(embedding)
                if embedding_norm > 0:
                    embedding = embedding / embedding_norm

                embeddings.append(embedding)

            except Exception as e:
                raise ValueError(f"Failed to load image {image_path}: {e}")

    logger.info(f"Computed fake embeddings for {len(embeddings)} images")
    return embeddings


def compute_text_embeddings(
    captions: list[str],
    batch_size: int = 32,
    model_name: str = "fake-embedding-for-testing",
) -> list[np.ndarray]:
    """
    Compute fake embeddings for a list of text captions.

    This function provides a fast, deterministic alternative to real text embedding models for testing.
    It computes 6-dimensional embeddings based on simple text statistics.

    Returns:
        List of 1D numpy arrays, one embedding vector per input caption
    """
    # Handle empty input
    if not captions:
        logger.info("No captions provided, returning empty list")
        return []

    logger.info(f"Computing fake text embeddings for {len(captions)} captions")

    embeddings = []

    # Process captions in batches (for API compatibility)
    for i in range(0, len(captions), batch_size):
        batch_captions = captions[i : i + batch_size]

        # Process each caption in the batch
        for caption in batch_captions:
            # Calculate simple text statistics
            text_length = len(caption)
            word_count = len(caption.split())
            char_diversity = len(set(caption.lower()))
            avg_word_length = text_length / max(word_count, 1)
            vowel_count = sum(1 for c in caption.lower() if c in "aeiou")
            consonant_count = sum(
                1 for c in caption.lower() if c.isalpha() and c not in "aeiou"
            )

            # Create 6-dimensional embedding from text statistics
            embedding = np.array(
                [
                    text_length / 100.0,  # Normalize text length
                    word_count / 20.0,  # Normalize word count
                    char_diversity / 26.0,  # Normalize character diversity
                    avg_word_length / 10.0,  # Normalize average word length
                    vowel_count / 50.0,  # Normalize vowel count
                    consonant_count / 50.0,  # Normalize consonant count
                ],
                dtype=np.float32,
            )

            # Normalize the embedding to unit length (to match SIGLIP2 behavior)
            embedding_norm = np.linalg.norm(embedding)
            if embedding_norm > 0:
                embedding = embedding / embedding_norm

            embeddings.append(embedding)

    logger.info(f"Computed fake text embeddings for {len(embeddings)} captions")
    return embeddings


def generate_captions_from_image_embeddings(
    image_embeddings: list[np.ndarray], batch_size: int = 8, max_length: int = 50
) -> list[str]:
    """
    Generate fake captions from fake image embeddings.

    Args:
        image_embeddings: List of image embedding vectors
        batch_size: Number of embeddings to process in each batch (for consistency)
        max_length: Maximum length of generated captions (for future use)

    Returns:
        List of generated caption strings
    """
    # Handle empty input
    if not image_embeddings:
        logger.info("No embeddings provided, returning empty list")
        return []

    logger.info(f"Generating fake captions for {len(image_embeddings)} embeddings")

    captions = []

    for i in range(0, len(image_embeddings), batch_size):
        batch_embeddings = image_embeddings[i : i + batch_size]

        # Generate captions based on embedding characteristics
        for embedding in batch_embeddings:
            # For fake embeddings, we know the structure: [R_mean, G_mean, B_mean, R_std, G_std, B_std]
            r_mean, g_mean, b_mean, r_std, g_std, b_std = embedding

            # Generate captions based on color characteristics
            if r_mean > g_mean and r_mean > b_mean:
                if r_std > 0.3:
                    captions.append("An image with varied red tones and high contrast")
                else:
                    captions.append("An image dominated by red colors")
            elif g_mean > r_mean and g_mean > b_mean:
                if g_std > 0.3:
                    captions.append(
                        "An image with varied green tones and natural elements"
                    )
                else:
                    captions.append("An image dominated by green colors")
            elif b_mean > r_mean and b_mean > g_mean:
                if b_std > 0.3:
                    captions.append("An image with varied blue tones like sky or water")
                else:
                    captions.append("An image dominated by blue colors")
            else:
                # Balanced colors
                overall_std = (r_std + g_std + b_std) / 3
                if overall_std > 0.25:
                    captions.append(
                        "A colorful image with balanced tones and high contrast"
                    )
                else:
                    captions.append("A balanced image with uniform lighting")

    logger.info(f"Generated {len(captions)} fake captions")
    return captions
