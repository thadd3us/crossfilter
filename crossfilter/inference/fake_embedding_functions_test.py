"""
Tests for fake embedding functions.
"""

import logging
from pathlib import Path

import numpy as np
import pytest
from syrupy import SnapshotAssertion

from crossfilter.inference.fake_embedding_functions import (
    FakeEmbedder,
)

logger = logging.getLogger(__name__)


@pytest.fixture
def embedder() -> FakeEmbedder:
    """Create FakeEmbedder instance for testing."""
    return FakeEmbedder()


@pytest.fixture
def test_image_paths() -> list[Path]:
    """Get paths to all test photos."""
    test_photos_dir = Path(__file__).parent.parent.parent / "test_data" / "test_photos"

    # Get all jpg files in the test_photos directory
    image_paths = sorted(test_photos_dir.glob("*.jpg"))

    # Verify we have the expected test images
    assert len(image_paths) == 10, f"Expected 10 test images, found {len(image_paths)}"

    # Verify all files exist
    for path in image_paths:
        assert path.exists(), f"Test image not found: {path}"

    return image_paths


@pytest.fixture
def test_captions() -> list[str]:
    """Sample captions for testing text embeddings."""
    return [
        "A beautiful golden gate bridge in San Francisco",
        "Planet Earth as seen from space showing blue oceans and white clouds",
        "A peaceful forest path with tall trees on both sides",
        "Stunning mountain landscape with snow-capped peaks",
        "Modern city skyline with tall skyscrapers at sunset",
        "Close-up view of colorful flowers in nature",
        "Busy street scene with people and vehicles",
        "Abstract geometric pattern with vibrant colors",
        "Urban architectural details and building textures",
    ]


def test_compute_image_embeddings(
    test_image_paths: list[Path], embedder: FakeEmbedder, snapshot: SnapshotAssertion
) -> None:
    """Test fake image embedding computation."""
    embeddings = embedder.compute_image_embeddings(test_image_paths)

    # Verify we got the expected number of embeddings
    assert len(embeddings) == len(test_image_paths)

    # Verify all embeddings are numpy arrays with expected properties
    for i, embedding in enumerate(embeddings):
        assert isinstance(embedding, np.ndarray), f"Embedding {i} is not a numpy array"
        assert embedding.ndim == 1, f"Embedding {i} is not 1D"
        assert (
            embedding.dtype == np.float32
        ), f"Embedding {i} has wrong dtype: {embedding.dtype}"
        assert (
            len(embedding) == 6
        ), f"Embedding {i} should have 6 dimensions, got {len(embedding)}"

        # Check that embedding is normalized (should have norm close to 1.0)
        norm = np.linalg.norm(embedding)
        assert 0.95 <= norm <= 1.05, f"Embedding {i} norm {norm} is not close to 1.0"

        # Check that values are reasonable (RGB stats should be in [0, 1] range before normalization)
        # After normalization, values can be outside this range, but should be finite
        assert np.all(
            np.isfinite(embedding)
        ), f"Embedding {i} contains non-finite values"

    # Create snapshot data with metadata
    snapshot_data = {
        "num_embeddings": len(embeddings),
        "embedding_dimension": len(embeddings[0]),
        "embedding_dtype": str(embeddings[0].dtype),
        "embedding_shapes": [embedding.shape for embedding in embeddings],
        "embedding_norms": [
            round(float(np.linalg.norm(embedding)), 4) for embedding in embeddings
        ],
        # Store all values of each embedding for regression testing (since they're only 6-dimensional)
        "embedding_values": [np.round(embedding, 4).tolist() for embedding in embeddings],
        "image_filenames": [path.name for path in test_image_paths],
    }

    assert snapshot_data == snapshot


def test_compute_text_embeddings(
    test_captions: list[str], embedder: FakeEmbedder, snapshot: SnapshotAssertion
) -> None:
    """Test fake text embedding computation for captions."""
    embeddings = embedder.compute_text_embeddings(test_captions)

    # Verify we got the expected number of embeddings
    assert len(embeddings) == len(test_captions)

    # Verify all embeddings are numpy arrays with expected properties
    for i, embedding in enumerate(embeddings):
        assert isinstance(
            embedding, np.ndarray
        ), f"Text embedding {i} is not a numpy array"
        assert embedding.ndim == 1, f"Text embedding {i} is not 1D"
        assert embedding.dtype == np.float32, f"Text embedding {i} has wrong dtype"
        assert (
            len(embedding) == 6
        ), f"Text embedding {i} should have 6 dimensions, got {len(embedding)}"

        # Check that embedding is normalized
        norm = np.linalg.norm(embedding)
        assert (
            0.95 <= norm <= 1.05
        ), f"Text embedding {i} norm {norm} is not close to 1.0"

        # Check for finite values
        assert np.all(
            np.isfinite(embedding)
        ), f"Text embedding {i} contains non-finite values"

    # Create snapshot data
    snapshot_data = {
        "num_embeddings": len(embeddings),
        "embedding_dimension": len(embeddings[0]),
        "embedding_dtype": str(embeddings[0].dtype),
        "embedding_norms": [
            round(float(np.linalg.norm(embedding)), 4) for embedding in embeddings
        ],
        # Store all values of each embedding
        "embedding_values": [
            [round(float(val), 4) for val in embedding.tolist()]
            for embedding in embeddings
        ],
        "input_captions": test_captions,
    }

    assert snapshot_data == snapshot


def test_generate_captions_from_image_embeddings(
    test_image_paths: list[Path], embedder: FakeEmbedder, snapshot: SnapshotAssertion
) -> None:
    """Test fake caption generation from fake image embeddings."""
    # First compute fake image embeddings
    image_embeddings = embedder.compute_image_embeddings(test_image_paths)

    # Generate captions from embeddings
    captions = embedder.generate_captions_from_image_embeddings(image_embeddings)

    # Verify we got the expected number of captions
    assert len(captions) == len(image_embeddings)

    # Verify all captions are strings
    for i, caption in enumerate(captions):
        assert isinstance(caption, str), f"Caption {i} is not a string"
        assert len(caption) > 0, f"Caption {i} is empty"

    # Create snapshot data
    snapshot_data = {
        "num_captions": len(captions),
        "captions": captions,
        "image_filenames": [path.name for path in test_image_paths],
        "caption_lengths": [len(caption) for caption in captions],
    }

    assert snapshot_data == snapshot




def test_error_handling_missing_image() -> None:
    """Test error handling for missing image files."""
    missing_path = Path("/nonexistent/image.jpg")
    embedder = FakeEmbedder()

    with pytest.raises(FileNotFoundError, match="Image not found"):
        embedder.compute_image_embeddings([missing_path])


def test_error_handling_invalid_image(tmp_path: Path) -> None:
    """Test error handling for invalid image files."""
    # Create a non-image file
    invalid_file = tmp_path / "not_an_image.jpg"
    invalid_file.write_text("This is not an image")
    embedder = FakeEmbedder()

    with pytest.raises(ValueError, match="Failed to load image"):
        embedder.compute_image_embeddings([invalid_file])


def test_empty_inputs() -> None:
    """Test handling of empty input lists."""
    embedder = FakeEmbedder()
    
    # Empty image list
    image_embeddings = embedder.compute_image_embeddings([])
    assert image_embeddings == []

    # Empty text list
    text_embeddings = embedder.compute_text_embeddings([])
    assert text_embeddings == []

    # Empty embeddings for caption generation
    captions = embedder.generate_captions_from_image_embeddings([])
    assert captions == []
