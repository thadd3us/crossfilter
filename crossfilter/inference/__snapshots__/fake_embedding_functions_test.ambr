# serializer version: 1
# name: test_compute_image_embeddings
  dict({
    'embedding_dimension': 6,
    'embedding_dtype': 'float32',
    'embedding_norms': list([
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'embedding_shapes': list([
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
    ]),
    'embedding_values': list([
      list([
        0.5246000289916992,
        0.5206000208854675,
        0.5292999744415283,
        0.23010000586509705,
        0.2304999977350235,
        0.2599000036716461,
      ]),
      list([
        0.4781999886035919,
        0.5077000260353088,
        0.5196999907493591,
        0.26570001244544983,
        0.28870001435279846,
        0.2994999885559082,
      ]),
      list([
        0.39800000190734863,
        0.4077000021934509,
        0.46720001101493835,
        0.37880000472068787,
        0.38040000200271606,
        0.41110000014305115,
      ]),
      list([
        0.4047999978065491,
        0.42239999771118164,
        0.42089998722076416,
        0.4004000127315521,
        0.40059998631477356,
        0.39980000257492065,
      ]),
      list([
        0.4645000100135803,
        0.4645000100135803,
        0.4645000100135803,
        0.34279999136924744,
        0.34279999136924744,
        0.34279999136924744,
      ]),
      list([
        0.4846000075340271,
        0.48089998960494995,
        0.4765999913215637,
        0.3197000026702881,
        0.3197999894618988,
        0.3197000026702881,
      ]),
      list([
        0.26010000705718994,
        0.21299999952316284,
        0.6151999831199646,
        0.45719999074935913,
        0.39890000224113464,
        0.37459999322891235,
      ]),
      list([
        0.5964000225067139,
        0.5400999784469604,
        0.4171999990940094,
        0.2621000111103058,
        0.23899999260902405,
        0.2296999990940094,
      ]),
      list([
        0.5022000074386597,
        0.5041999816894531,
        0.5051000118255615,
        0.28439998626708984,
        0.2831000089645386,
        0.2782999873161316,
      ]),
      list([
        0.5056999921798706,
        0.5056999921798706,
        0.5056999921798706,
        0.2784999907016754,
        0.2784999907016754,
        0.2784999907016754,
      ]),
    ]),
    'image_filenames': list([
      '00_munich_rathaus.jpg',
      '01_golden_gate_bridge.jpg',
      '02_earth_from_space.jpg',
      '03_backlit_man_looking_out.jpg',
      '04_mountain_view.jpg',
      '05_astronaut_on_moon.jpg',
      '06_fireworks_on_blue_sunset.jpg',
      '07_paper_bundle_and_pen.jpg',
      '08_herman_hesse.jpg',
      '09_martin_luther_king_jr.jpg',
    ]),
    'num_embeddings': 10,
  })
# ---
# name: test_compute_image_embeddings_all_at_once
  dict({
    'embedding_dimension': 6,
    'embedding_values': list([
      list([
        0.5246466398239136,
        0.5206447839736938,
        0.5292510986328125,
        0.2300640493631363,
        0.23046532273292542,
        0.2598549425601959,
      ]),
      list([
        0.4781765937805176,
        0.5076568126678467,
        0.519658625125885,
        0.26565104722976685,
        0.2886870205402374,
        0.2994595766067505,
      ]),
      list([
        0.3979630470275879,
        0.4076855182647705,
        0.4672101140022278,
        0.3788158595561981,
        0.38035187125205994,
        0.411051869392395,
      ]),
      list([
        0.40484437346458435,
        0.42235270142555237,
        0.4208824932575226,
        0.4003521800041199,
        0.4005855619907379,
        0.3997829854488373,
      ]),
      list([
        0.4645412862300873,
        0.4645412862300873,
        0.4645412862300873,
        0.34283339977264404,
        0.34283339977264404,
        0.34283339977264404,
      ]),
      list([
        0.4846378564834595,
        0.4809095561504364,
        0.47660142183303833,
        0.31973743438720703,
        0.3197622299194336,
        0.31972360610961914,
      ]),
      list([
        0.2601301074028015,
        0.21302247047424316,
        0.6152127981185913,
        0.457185834646225,
        0.39890074729919434,
        0.3746016025543213,
      ]),
      list([
        0.5963801145553589,
        0.5401008129119873,
        0.41721293330192566,
        0.2620631158351898,
        0.2389506846666336,
        0.2297404259443283,
      ]),
      list([
        0.502216100692749,
        0.5042033791542053,
        0.5050593018531799,
        0.28439807891845703,
        0.28313541412353516,
        0.2782537341117859,
      ]),
      list([
        0.5057275295257568,
        0.5057275295257568,
        0.5057275295257568,
        0.27851927280426025,
        0.27851927280426025,
        0.27851927280426025,
      ]),
    ]),
    'num_embeddings': 10,
  })
# ---
# name: test_compute_image_embeddings_single_batch
  dict({
    'embedding_dimension': 6,
    'embedding_dtype': 'float32',
    'embedding_norms': list([
      1.0,
      1.0,
      1.0,
      0.9999999403953552,
      1.0,
      0.9999998807907104,
      1.0,
      0.9999999403953552,
      1.0,
      1.0,
    ]),
    'embedding_shapes': list([
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
      tuple(
        6,
      ),
    ]),
    'embedding_values': list([
      list([
        0.5246466398239136,
        0.5206447839736938,
        0.5292510986328125,
        0.2300640493631363,
        0.23046532273292542,
        0.2598549425601959,
      ]),
      list([
        0.4781765937805176,
        0.5076568126678467,
        0.519658625125885,
        0.26565104722976685,
        0.2886870205402374,
        0.2994595766067505,
      ]),
      list([
        0.3979630470275879,
        0.4076855182647705,
        0.4672101140022278,
        0.3788158595561981,
        0.38035187125205994,
        0.411051869392395,
      ]),
      list([
        0.40484437346458435,
        0.42235270142555237,
        0.4208824932575226,
        0.4003521800041199,
        0.4005855619907379,
        0.3997829854488373,
      ]),
      list([
        0.4645412862300873,
        0.4645412862300873,
        0.4645412862300873,
        0.34283339977264404,
        0.34283339977264404,
        0.34283339977264404,
      ]),
      list([
        0.4846378564834595,
        0.4809095561504364,
        0.47660142183303833,
        0.31973743438720703,
        0.3197622299194336,
        0.31972360610961914,
      ]),
      list([
        0.2601301074028015,
        0.21302247047424316,
        0.6152127981185913,
        0.457185834646225,
        0.39890074729919434,
        0.3746016025543213,
      ]),
      list([
        0.5963801145553589,
        0.5401008129119873,
        0.41721293330192566,
        0.2620631158351898,
        0.2389506846666336,
        0.2297404259443283,
      ]),
      list([
        0.502216100692749,
        0.5042033791542053,
        0.5050593018531799,
        0.28439807891845703,
        0.28313541412353516,
        0.2782537341117859,
      ]),
      list([
        0.5057275295257568,
        0.5057275295257568,
        0.5057275295257568,
        0.27851927280426025,
        0.27851927280426025,
        0.27851927280426025,
      ]),
    ]),
    'image_filenames': list([
      '00_munich_rathaus.jpg',
      '01_golden_gate_bridge.jpg',
      '02_earth_from_space.jpg',
      '03_backlit_man_looking_out.jpg',
      '04_mountain_view.jpg',
      '05_astronaut_on_moon.jpg',
      '06_fireworks_on_blue_sunset.jpg',
      '07_paper_bundle_and_pen.jpg',
      '08_herman_hesse.jpg',
      '09_martin_luther_king_jr.jpg',
    ]),
    'num_embeddings': 10,
  })
# ---
# name: test_compute_image_embeddings_small_batches
  dict({
    'batch_size_used': 3,
    'embedding_dimension': 6,
    'embedding_values': list([
      list([
        0.4825541377067566,
        0.5013473033905029,
        0.5116933584213257,
        0.2700468599796295,
        0.2948189675807953,
        0.3067876696586609,
      ]),
      list([
        0.3979630470275879,
        0.4076855182647705,
        0.4672101140022278,
        0.3788158595561981,
        0.38035187125205994,
        0.411051869392395,
      ]),
      list([
        0.40621417760849,
        0.4237518608570099,
        0.42235177755355835,
        0.39890703558921814,
        0.39906004071235657,
        0.39833104610443115,
      ]),
      list([
        0.4694303870201111,
        0.4694303870201111,
        0.4694303870201111,
        0.3361077904701233,
        0.3361077904701233,
        0.3361077904701233,
      ]),
      list([
        0.520820140838623,
        0.5110766887664795,
        0.43892204761505127,
        0.31178703904151917,
        0.31369560956954956,
        0.28156420588493347,
      ]),
      list([
        0.3058488965034485,
        0.25643190741539,
        0.5879448056221008,
        0.45111361145973206,
        0.3998323082923889,
        0.36283689737319946,
      ]),
      list([
        0.6005426645278931,
        0.5368449091911316,
        0.40097934007644653,
        0.27653586864471436,
        0.24665617942810059,
        0.2303265631198883,
      ]),
      list([
        0.4417613744735718,
        0.44894421100616455,
        0.4267692267894745,
        0.4033930003643036,
        0.38190412521362305,
        0.3355403542518616,
      ]),
      list([
        0.5120797157287598,
        0.4622848331928253,
        0.4283306300640106,
        0.3522045314311981,
        0.3466721475124359,
        0.31043586134910583,
      ]),
    ]),
    'num_embeddings': 9,
  })
# ---
# name: test_compute_text_embeddings
  dict({
    'embedding_dimension': 6,
    'embedding_dtype': 'float32',
    'embedding_norms': list([
      0.9999999403953552,
      1.0,
      1.0,
      0.9999999403953552,
      1.0,
      0.9999999403953552,
      1.0,
      0.9999999403953552,
      1.0,
    ]),
    'embedding_values': list([
      list([
        0.39276543259620667,
        0.3342684507369995,
        0.5142591595649719,
        0.49095675349235535,
        0.28412818908691406,
        0.38440871238708496,
      ]),
      list([
        0.4372265338897705,
        0.38578811287879944,
        0.4946001470088959,
        0.3643554151058197,
        0.2829112708568573,
        0.4500861167907715,
      ]),
      list([
        0.40310201048851013,
        0.38759809732437134,
        0.5366743206977844,
        0.40310201048851013,
        0.24806277453899384,
        0.4186059534549713,
      ]),
      list([
        0.35400786995887756,
        0.21240472793579102,
        0.517396092414856,
        0.5900130867958069,
        0.21240472793579102,
        0.4106491208076477,
      ]),
      list([
        0.37723690271377563,
        0.2958720922470093,
        0.5405355095863342,
        0.47154611349105835,
        0.17752324044704437,
        0.47339531779289246,
      ]),
      list([
        0.3662179112434387,
        0.2980843484401703,
        0.5568608641624451,
        0.5231684446334839,
        0.2555008828639984,
        0.35770121216773987,
      ]),
      list([
        0.34616023302078247,
        0.28846684098243713,
        0.6022934317588806,
        0.4945146441459656,
        0.21428966522216797,
        0.3791278898715973,
      ]),
      list([
        0.3477404713630676,
        0.226787269115448,
        0.5233551859855652,
        0.5795674324035645,
        0.1965489387512207,
        0.4233362078666687,
      ]),
      list([
        0.3669673204421997,
        0.22467388212680817,
        0.4608694911003113,
        0.6116122007369995,
        0.2546303868293762,
        0.40441298484802246,
      ]),
    ]),
    'input_captions': list([
      'A beautiful golden gate bridge in San Francisco',
      'Planet Earth as seen from space showing blue oceans and white clouds',
      'A peaceful forest path with tall trees on both sides',
      'Stunning mountain landscape with snow-capped peaks',
      'Modern city skyline with tall skyscrapers at sunset',
      'Close-up view of colorful flowers in nature',
      'Busy street scene with people and vehicles',
      'Abstract geometric pattern with vibrant colors',
      'Urban architectural details and building textures',
    ]),
    'num_embeddings': 9,
  })
# ---
# name: test_generate_captions_from_image_embeddings
  dict({
    'caption_lengths': list([
      33,
      33,
      49,
      53,
      54,
      48,
      49,
      32,
      33,
      54,
    ]),
    'captions': list([
      'An image dominated by blue colors',
      'An image dominated by blue colors',
      'An image with varied blue tones like sky or water',
      'An image with varied green tones and natural elements',
      'A colorful image with balanced tones and high contrast',
      'An image with varied red tones and high contrast',
      'An image with varied blue tones like sky or water',
      'An image dominated by red colors',
      'An image dominated by blue colors',
      'A colorful image with balanced tones and high contrast',
    ]),
    'image_filenames': list([
      '00_munich_rathaus.jpg',
      '01_golden_gate_bridge.jpg',
      '02_earth_from_space.jpg',
      '03_backlit_man_looking_out.jpg',
      '04_mountain_view.jpg',
      '05_astronaut_on_moon.jpg',
      '06_fireworks_on_blue_sunset.jpg',
      '07_paper_bundle_and_pen.jpg',
      '08_herman_hesse.jpg',
      '09_martin_luther_king_jr.jpg',
    ]),
    'num_captions': 10,
  })
# ---
